File: frontend/src/types/api/response.ts
/**
 * API レスポンスの統一的な型定義
 * 
 * バックエンドのPydanticスキーマと対応する
 * TypeScriptの型定義を提供します。
 */

/**
 * 成功レスポンスの基本型
 */
export interface SuccessResponse<T> {
  success: true;
  data: T;
  message?: string;
  timestamp: string;
}

/**
 * エラーの詳細情報
 */
export interface ErrorDetail {
  field?: string;
  reason: string;
  value?: any;
}

/**
 * エラー情報
 */
export interface ErrorInfo {
  code: string;
  message: string;
  timestamp: string;
  details?: ErrorDetail[];
  request_id?: string;
  trace_id?: string;
}

/**
 * エラーレスポンス
 */
export interface ErrorResponse {
  error: ErrorInfo;
  status_code: number;
}

/**
 * ページネーション情報
 */
export interface PaginationMeta {
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

/**
 * ページネーション付きレスポンス
 */
export interface PaginatedResponse<T> {
  success: true;
  data: T[];
  meta: PaginationMeta;
  message?: string;
  timestamp: string;
}

/**
 * 一括操作の結果
 */
export interface BulkOperationResult {
  succeeded: number;
  failed: number;
  errors: ErrorDetail[];
}

/**
 * 一括操作レスポンス
 */
export interface BulkOperationResponse {
  success: boolean;
  result: BulkOperationResult;
  message: string;
  timestamp: string;
}

/**
 * ヘルスチェックレスポンス
 */
export interface HealthCheckResponse {
  status: string;
  timestamp: string;
  version?: string;
  services?: Record<string, any>;
}

/**
 * メッセージレスポンス
 */
export interface MessageResponse {
  success: boolean;
  message: string;
  timestamp: string;
}

/**
 * ステータスレスポンス
 */
export interface StatusResponse {
  status: string;
  details?: Record<string, any>;
  timestamp: string;
}

/**
 * APIレスポンスの共用型
 */
export type ApiResponse<T> = SuccessResponse<T> | ErrorResponse;

/**
 * ページネーション付きAPIレスポンスの共用型
 */
export type PaginatedApiResponse<T> = PaginatedResponse<T> | ErrorResponse;

/**
 * 型ガード: 成功レスポンスかどうかを判定
 */
export function isSuccessResponse<T>(
  response: ApiResponse<T>
): response is SuccessResponse<T> {
  return 'success' in response && response.success === true;
}

/**
 * 型ガード: エラーレスポンスかどうかを判定
 */
export function isErrorResponse(
  response: ApiResponse<any>
): response is ErrorResponse {
  return 'error' in response && 'status_code' in response;
}

/**
 * 型ガード: ページネーション付きレスポンスかどうかを判定
 */
export function isPaginatedResponse<T>(
  response: any
): response is PaginatedResponse<T> {
  return (
    'success' in response &&
    response.success === true &&
    'meta' in response &&
    Array.isArray(response.data)
  );
}

/**
 * エラーコードの定数
 */
export const ERROR_CODES = {
  // クライアントエラー
  VALIDATION_ERROR: 'VALIDATION_ERROR',
  UNAUTHORIZED: 'UNAUTHORIZED',
  FORBIDDEN: 'FORBIDDEN',
  NOT_FOUND: 'NOT_FOUND',
  CONFLICT: 'CONFLICT',
  TOO_MANY_REQUESTS: 'TOO_MANY_REQUESTS',
  
  // サーバーエラー
  INTERNAL_SERVER_ERROR: 'INTERNAL_SERVER_ERROR',
  BAD_GATEWAY: 'BAD_GATEWAY',
  SERVICE_UNAVAILABLE: 'SERVICE_UNAVAILABLE',
  
  // ネットワーク・その他
  NETWORK_ERROR: 'NETWORK_ERROR',
  TIMEOUT: 'TIMEOUT',
  UNKNOWN_ERROR: 'UNKNOWN_ERROR'
} as const;

export type ErrorCode = typeof ERROR_CODES[keyof typeof ERROR_CODES];

/**
 * APIリクエストのオプション
 */
export interface ApiRequestOptions {
  timeout?: number;
  retry?: {
    count: number;
    delay: number;
  };
  signal?: AbortSignal;
}

/**
 * フィルター条件の基本型
 */
export interface BaseFilters {
  [key: string]: any;
}

/**
 * ソート条件の基本型
 */
export interface SortOptions {
  field: string;
  order: 'asc' | 'desc';
}

/**
 * リスト取得のパラメータ
 */
export interface ListParams<TFilters extends BaseFilters = BaseFilters> {
  page?: number;
  per_page?: number;
  filters?: TFilters;
  sort?: SortOptions;
}

/**
 * 検索パラメータ
 */
export interface SearchParams extends ListParams {
  query: string;
}

/**
 * 日付範囲フィルター
 */
export interface DateRangeFilter {
  start?: string;
  end?: string;
}

/**
 * 数値範囲フィルター
 */
export interface NumberRangeFilter {
  min?: number;
  max?: number;
}

/**
 * APIクライアントの基本インターフェース
 */
export interface ApiClient {
  get<T>(url: string, options?: ApiRequestOptions): Promise<T>;
  post<T>(url: string, data?: any, options?: ApiRequestOptions): Promise<T>;
  put<T>(url: string, data?: any, options?: ApiRequestOptions): Promise<T>;
  patch<T>(url: string, data?: any, options?: ApiRequestOptions): Promise<T>;
  delete<T>(url: string, options?: ApiRequestOptions): Promise<T>;
}

File: backend/app/schemas/response.py
"""
API レスポンスの統一的なスキーマ定義

このモジュールは、API全体で使用される統一的なレスポンス
フォーマットのPydanticスキーマを定義します。
"""

from typing import TypeVar, Generic, Optional, List, Any
from datetime import datetime
from pydantic import BaseModel, Field
from pydantic.generics import GenericModel


# 汎用的な型変数
DataT = TypeVar('DataT')


class SuccessResponse(GenericModel, Generic[DataT]):
    """成功レスポンスの基本スキーマ"""
    
    success: bool = Field(True, description="成功フラグ")
    data: DataT = Field(..., description="レスポンスデータ")
    message: Optional[str] = Field(None, description="成功メッセージ")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat(),
        description="レスポンス生成時刻"
    )


class ErrorDetail(BaseModel):
    """エラーの詳細情報"""
    field: Optional[str] = Field(None, description="エラーが発生したフィールド")
    reason: str = Field(..., description="エラーの理由")
    value: Optional[Any] = Field(None, description="問題のある値")


class ErrorResponse(BaseModel):
    """エラーレスポンスのスキーマ"""
    
    error: dict = Field(..., description="エラー情報")
    status_code: int = Field(..., description="HTTPステータスコード")
    
    class Config:
        schema_extra = {
            "example": {
                "error": {
                    "code": "VALIDATION_ERROR",
                    "message": "入力値が無効です",
                    "timestamp": "2024-01-01T00:00:00Z",
                    "details": [
                        {
                            "field": "email",
                            "reason": "有効なメールアドレスではありません",
                            "value": "invalid-email"
                        }
                    ]
                },
                "status_code": 422
            }
        }


class PaginationMeta(BaseModel):
    """ページネーション情報"""
    total: int = Field(..., ge=0, description="総件数")
    page: int = Field(..., ge=1, description="現在のページ番号")
    per_page: int = Field(..., ge=1, le=100, description="1ページあたりの件数")
    total_pages: int = Field(..., ge=0, description="総ページ数")
    has_next: bool = Field(..., description="次ページの有無")
    has_prev: bool = Field(..., description="前ページの有無")


class PaginatedResponse(GenericModel, Generic[DataT]):
    """ページネーション付きレスポンス"""
    
    success: bool = Field(True, description="成功フラグ")
    data: List[DataT] = Field(..., description="データリスト")
    meta: PaginationMeta = Field(..., description="ページネーション情報")
    message: Optional[str] = Field(None, description="メッセージ")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat(),
        description="レスポンス生成時刻"
    )


class BulkOperationResult(BaseModel):
    """一括操作の結果"""
    
    succeeded: int = Field(0, ge=0, description="成功件数")
    failed: int = Field(0, ge=0, description="失敗件数")
    errors: List[ErrorDetail] = Field(
        default_factory=list,
        description="エラー詳細リスト"
    )


class BulkOperationResponse(BaseModel):
    """一括操作レスポンス"""
    
    success: bool = Field(..., description="全体の成功フラグ")
    result: BulkOperationResult = Field(..., description="操作結果")
    message: str = Field(..., description="結果メッセージ")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat(),
        description="レスポンス生成時刻"
    )


class HealthCheckResponse(BaseModel):
    """ヘルスチェックレスポンス"""
    
    status: str = Field(..., description="サービスステータス")
    timestamp: str = Field(..., description="チェック時刻")
    version: Optional[str] = Field(None, description="APIバージョン")
    services: Optional[dict] = Field(None, description="依存サービスの状態")


class MessageResponse(BaseModel):
    """シンプルなメッセージレスポンス"""
    
    success: bool = Field(..., description="成功フラグ")
    message: str = Field(..., description="メッセージ")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat(),
        description="レスポンス生成時刻"
    )


class StatusResponse(BaseModel):
    """ステータス確認レスポンス"""
    
    status: str = Field(..., description="ステータス")
    details: Optional[dict] = Field(None, description="詳細情報")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat(),
        description="レスポンス生成時刻"
    )


# 便利な型エイリアス
StringListResponse = SuccessResponse[List[str]]
DictResponse = SuccessResponse[dict]